solidity/contracts/LinkEx.sol

```diff
@@ -18,15 +37,46 @@ contract LinkEx is LinkExInterface {
     return historicRate;
   }
 
-  function update(uint256 _rate) external {
+  function removeOracle(address _oracle) external onlyOwner {
+    setFulfillmentPermission(_oracle, false);
+    delete authorizedNodes[_oracle];
+    for (uint i = 0; i < oracles.length; i++) {
+      if (oracles[i] == _oracle) {
+        delete rates[oracles[i]];
+        delete oracles[i];
```

Comment:
Does this reorder the list? I think it just leaves an empty element in that index, so when we decrement the length in the next line, it leaves an empty spot. If I'm right, it would be exposed with a test involving three oracles, and oracle number two gets deleted.

One solution might be to get the last oracle in the list, delete it from there and copy it to the previous location, but that may well have some edge cases to it.

Reply:
Okay good. We shouldn't be touching the length because that will effect every element of the array after it. The test was only deleting the last element so there didn't appear to be a problem, but when updating that to delete the first element is where everything goes wrong. This is now corrected by simply leaving the length as-is on removal of an oracle.

Reply:
I don't think it's that simple, we now have an empty oracle in the list, ie `oracles[i] == adress(0)`, right? I think we need to decrement the count. Then, reorder the oracle array, except if the oracle deleted is the last oracle in the list.

Reply:
Turns out it was simple, just not as efficient as it could be. We did have a test already for deleting an oracle at the beginning of the array and running the average again. What happened was the update loop would hit that empty place in the array, see that blocknumber (being 0) was older than 25 blocks, and not count it towards the average. Now upon removing an oracle, if the target oracle isn't the last in the array, the last oracle in the array will take the place of the deleted oracle, and we'll decrement the array length. This seems to save ~1000 gas per-update in my small-scale test.

Reply:
We'd also get a gas refund if we also deleted the last spot in that array, right?

Reply:
I had to refactor that `removeOracle` method a bit to get some more gas savings out of it, but yes:
Start when deleting from middle of array: 42324
After adding the line to delete the last oracle element: 45242
After adding the line to calculate size once: 44940
After adding return values to end early once removed: 44434
After removing the duplicate setFulfillmentPermission method call: 41738
After using the calculated size for the new length: 41584

Reply:
Actually I just came across [this](https://ethereum.stackexchange.com/a/39311/37908) answer. I'll do some tests to see if this holds true. If so, we might be able to save another ~3000 or so gas when removing an oracle.

Reply:
Based on that SO answer, can't we get rid of the `delete oracles[i];` on line 53? Seems like it is handled by 54-57.